### PR TITLE
first

### PR DESCRIPTION
**Pull request by voloshynamaryna11**

### CODE REVIEW

File: src/main/java/cinema/dao/impl/UserDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import cinema.util.HibernateUtil;
+import java.util.Optional;
+import org.hibernate.Session;
+import org.hibernate.Transaction;
+import org.hibernate.query.Query;
+
+@Dao
+public class UserDaoImpl implements UserDao {
+
+    @Override
```
Comment (dariiabutorina) [inline src/main/java/cinema/dao/impl/UserDaoImpl.java]:
```suggestion
public class UserDaoImpl implements UserDao {
    @Override
```

File: src/main/java/cinema/dao/impl/UserDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        }
+    }
+
+    @Override
+    public Optional<User> findByEmail(String email) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            Query<User> findByEmailQuery = session.createQuery("FROM User "
+                    + "WHERE email = :email", User.class);
+            findByEmailQuery.setParameter("email", email);
+            return Optional.ofNullable(findByEmailQuery.uniqueResult());
```
Comment (dariiabutorina) [inline src/main/java/cinema/dao/impl/UserDaoImpl.java]: There is a special method called **_uniqueResultOptional()_** :)

File: src/main/java/cinema/model/User.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        this.salt = salt;
+    }
+
+    @Override
+    public String toString() {
+        return "User{"
+                + "id=" + id
+                + ", email='" + email + '\''
+                + ", password='" + password + '\''
+                + ", salt=" + Arrays.toString(salt)
```
Comment (dariiabutorina) [inline src/main/java/cinema/model/User.java]: Let's not show a password and a salt

File: src/main/java/cinema/security/AuthenticationServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import cinema.lib.Service;
+import cinema.model.User;
+import cinema.service.UserService;
+import cinema.util.HashUtil;
+import java.util.Optional;
+
+@Service
+public class AuthenticationServiceImpl implements AuthenticationService {
+
+    @Inject
```
Comment (dariiabutorina) [inline src/main/java/cinema/security/AuthenticationServiceImpl.java]:
```suggestion
public class AuthenticationServiceImpl implements AuthenticationService {
    @Inject
```

File: src/main/java/cinema/security/AuthenticationServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    @Override
+    public User register(String email, String password) {
+        User user = new User();
+        user.setEmail(email);
+        byte[] salt = HashUtil.getSalt();
+        String saltedPassword = HashUtil.hashPassword(password, salt);
+        user.setPassword(saltedPassword);
+        user.setSalt(salt);
```
Comment (dariiabutorina) [inline src/main/java/cinema/security/AuthenticationServiceImpl.java]: All the business logic somehow related to initializing or setting the values to our fields must be on the **_service_** layer.
If somebody creates the object **_User_** not throughout the **_register_** method but using the **_userService.create()_** instead, we'll not have its password hashed, which might cause some errors in the future while working with the object or with the security itself.
So let's move the hashing logic to the **_service_** - **_create_** method.

File: src/main/java/cinema/service/impl/UserServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+import cinema.lib.Inject;
+import cinema.lib.Service;
+import cinema.model.User;
+import cinema.service.UserService;
+import java.util.Optional;
+
+@Service
+public class UserServiceImpl implements UserService {
+
+    @Inject
```
Comment (dariiabutorina) [inline src/main/java/cinema/service/impl/UserServiceImpl.java]:
```suggestion
public class UserServiceImpl implements UserService {
    @Inject
```

**Review by dariiabutorina — CHANGES_REQUESTED** [review]

**Comment by Sofasmile** [comment]
Для початку виправи коментарі @DariiaPikul

File: src/main/java/cinema/dao/impl/UserDaoImpl.java (hunk 1 above)
Comment (voloshynamaryna11) [reply src/main/java/cinema/dao/impl/UserDaoImpl.java]: done

File: src/main/java/cinema/dao/impl/UserDaoImpl.java (hunk 2 above)
Comment (voloshynamaryna11) [reply src/main/java/cinema/dao/impl/UserDaoImpl.java]: done

File: src/main/java/cinema/model/User.java (hunk 3 above)
Comment (voloshynamaryna11) [reply src/main/java/cinema/model/User.java]: done

File: src/main/java/cinema/security/AuthenticationServiceImpl.java (hunk 4 above)
Comment (voloshynamaryna11) [reply src/main/java/cinema/security/AuthenticationServiceImpl.java]: done

File: src/main/java/cinema/service/impl/UserServiceImpl.java (hunk 6 above)
Comment (voloshynamaryna11) [reply src/main/java/cinema/service/impl/UserServiceImpl.java]: done

File: src/main/java/cinema/dao/impl/UserDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+            session = HibernateUtil.getSessionFactory().openSession();
+            transaction = session.beginTransaction();
+            session.save(user);
+            transaction.commit();
+            return user;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcessingException("Can't insert user Entity with id = "
```
Comment (Sofasmile) [inline src/main/java/cinema/dao/impl/UserDaoImpl.java]: ми уже знаємо тут ід юзера?

File: src/main/java/cinema/dao/impl/UserDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+    @Override
+    public Optional<User> findByEmail(String email) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            Query<User> findByEmailQuery = session.createQuery("FROM User "
+                    + "WHERE email = :email", User.class);
+            findByEmailQuery.setParameter("email", email);
+            return findByEmailQuery.uniqueResultOptional();
+        } catch (Exception e) {
+            throw new RuntimeException("Can't find available account", e);
```
Comment (Sofasmile) [inline src/main/java/cinema/dao/impl/UserDaoImpl.java]: DataProcessingException 
і давай зробимо повідомлення більш інформативним

File: src/main/java/cinema/exceptions/AuthenticationException.java [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package cinema.exceptions;
+
+public class AuthenticationException extends RuntimeException {
```
Comment (Sofasmile) [inline src/main/java/cinema/exceptions/AuthenticationException.java]: https://mate-academy.github.io/jv-program-fulltime/03_hibernate/checklist/hibernate_05.html
9 пункт

File: src/main/java/cinema/util/HashUtil.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
+package cinema.util;
+
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+
+public class HashUtil {
+    private static final String hashAlgName = "SHA-512";
```
Comment (Sofasmile) [inline src/main/java/cinema/util/HashUtil.java]: https://mate-academy.github.io/style-guides/java/java.html#s5.2.4-constant-names

File: src/main/java/cinema/util/HashUtil.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+        StringBuilder hashedPassword = new StringBuilder();
+        try {
+            MessageDigest messageDigest = MessageDigest.getInstance(hashAlgName);
+            messageDigest.update(salt);
+            byte[] digest = messageDigest.digest(password.getBytes());
+            for (byte b : digest) {
+                hashedPassword.append(String.format("%02x", b));
+            }
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException("Impossible to encode your password");
```
Comment (Sofasmile) [inline src/main/java/cinema/util/HashUtil.java]:
```suggestion
            throw new RuntimeException("Impossible to encode your password", e);
```

File: src/main/java/cinema/dao/impl/UserDaoImpl.java (hunk 8 above)
Comment (voloshynamaryna11) [reply src/main/java/cinema/dao/impl/UserDaoImpl.java]: done

File: src/main/java/cinema/exceptions/AuthenticationException.java (hunk 9 above)
Comment (voloshynamaryna11) [reply src/main/java/cinema/exceptions/AuthenticationException.java]: done

File: src/main/java/cinema/util/HashUtil.java (hunk 10 above)
Comment (voloshynamaryna11) [reply src/main/java/cinema/util/HashUtil.java]: done

File: src/main/java/cinema/util/HashUtil.java (hunk 11 above)
Comment (voloshynamaryna11) [reply src/main/java/cinema/util/HashUtil.java]: done

File: src/main/java/cinema/dao/impl/UserDaoImpl.java (hunk 7 above)
Comment (voloshynamaryna11) [reply src/main/java/cinema/dao/impl/UserDaoImpl.java]: done

**Review by Sofasmile — APPROVED** [review]

**Pull request merged**